apps/anoma_node/lib/node/transaction/shard.ex

```diff
@@ -0,0 +1,767 @@
+  # - The heights of all write reservations between that value and target_height.
+  @spec find_essential_heights_below(height(), map()) :: MapSet.t(height())
+  defp find_essential_heights_below(target_height, key_height_map) do
+    # 1. Find the highest height h_val < target_height with a committed value
+    maybe_max_h_val =
+      key_height_map
+      |> Enum.filter(fn {h, details} ->
+        h < target_height and not is_nil(details.value)
+      end)
+      |> Enum.max_by(fn {h, _} -> h end, fn -> nil end)
+
+    case maybe_max_h_val do
+      nil ->
+        # No committed value below target_height. Find write reservations below target_height.
+        write_reservation_heights_below =
+          key_height_map
+          |> Enum.filter(fn {h, details} ->
+            h < target_height and details.write_reserved?
+          end)
+          # Keep only the heights
+          |> Enum.map(fn {h, _} -> h end)
+
+        MapSet.new(write_reservation_heights_below)
+
+      {h_val, _details} ->
+        # 2. Find all heights h_wr with write reservations between h_val and target_height
+        write_reservation_heights_between =
+          key_height_map
+          |> Enum.filter(fn {h, details} ->
+            h > h_val and h < target_height and details.write_reserved?
+          end)
+          # Keep only the heights
+          |> Enum.map(fn {h, _} -> h end)
+
+        # 3. Combine h_val and the intermediate write reservation heights
+        MapSet.new([h_val | write_reservation_heights_between])
+    end
+  end
+
+  # I am the garbage collection helper function.
+
+  # I perform garbage collection for a specific key based on the read watermark.
+  # I remove entries older than the watermark unless they are essential for resolving
+  # reads at or past the watermark height or at heights with active read reservations.
+  @spec gc_key(key(), height(), __MODULE__.t()) :: __MODULE__.t()
+  defp gc_key(key, read_watermark, state) do
+    case Map.get(state.kv, key) do
+      nil ->
+        # Key not present, nothing to GC
+        state
+
+      key_height_map ->
+        # 1. Identify heights with active read reservations
+        read_reservation_heights =
+          for {h, details} <- key_height_map, details.read_reserved?, do: h
+
+        read_reservation_heights_set = MapSet.new(read_reservation_heights)
+
+        # 2. Determine essential heights to keep below the read watermark
+        essential_below_watermark =
+          find_essential_heights_below(read_watermark, key_height_map)
+
+        # 3. Determine essential heights to keep below each active read reservation
+        essential_below_reservations =
+          Enum.reduce(read_reservation_heights, MapSet.new(), fn h_rr,
+                                                                 acc_set ->
+            MapSet.union(
+              acc_set,
+              find_essential_heights_below(h_rr, key_height_map)
+            )
+          end)
```

Comment:
```suggestion
          read_reservation_heights
          |> Enum.map(&find_essential_heights_below(&1, key_height_map))
          |> Enum.reduce(MapSet.union/2)
```

basically we can turn this into a clean map reduce, making the reduce properly associative and the map explicit, this makes the code easier to read and reason about